### DISCUSSION TITLE
ASP.NET Core Microsoft DI and how to create job/triggers only if they don't already exist.

### DISCUSSION THREAD

**Discussion by jafin** [quartznet/quartznet#969] [General] [answered]
I am attempting to bootstrap Quartz.net in an asp.net core 3.1 app
I am using the UseMicrosoftDependencyInjectionScopedJobFactory

The configuration overwrites existing jobs and schedules.
How best can I avoid this,  I'd attempting to have the startup code only create the job/trigger _IF_ it currently does not exist in the PersistentStore.   

With the below config, it is overwriting the existing job/trigger, which resets the last PREV_FIRE_TIME.

```c#
q.SchedulerId = "Scheduler-Core";
q.UseMicrosoftDependencyInjectionScopedJobFactory(options => { });
q.UseSimpleTypeLoader();
q.UsePersistentStore(config =>
 {
  config.UseSqlServer(Configuration.GetConnectionString("DataConnection"));
  config.UseJsonSerializer();
 });
```

I configure the Jobs 

```c#
     q.AddJob<EmailReportJob>(job =>
       job.WithIdentity(jobKey).WithDescription("Description").StoreDurably(true));

q.AddTrigger(t =>
        t.WithIdentity("Trigger", JobGroup)
        .ForJob("emailReportJob", JobGroup)
        .StartNow()
        .WithCronSchedule("0 48 14 ? * TUE *", x => x.WithMisfireHandlingInstructionFireAndProceed()));
```

**Comment by lahma** [comment, marked as answer]
Try adding following before `AddQuartz`:

```csharp
services.AddSingleton(new QuartzSchedulingOptions
{
    OverWriteExistingData = false,
    IgnoreDuplicates = true
});
```

Please note that the configuration infrastructure will be tuned a bit in 3.2 it might require some changes then. But I'll try to make sure that how to do this is also documented.

**Reply by jafin** [reply]
Perfect that works!   Thanks @lahma

For my simple case this outcome is good. 

For more advanced scenarios (multiple jobs with different overwrite conditions per job), It would be useful if on a per trigger/job you could set the overwriteExistingData value to allow more granular control.  Not sure if that is in scope for changes of 3.2

**Reply by lahma** [reply]
When things get complicated it might be just easier to get handle to IScheduler and add/update/remove with your own logic. The extensions methods are modeled with same logic as the XML file support (which is also an option here).

**Reply by jafin** [reply]
yep ok,  before your initial reply that's essentially the path I ended down.  (IScheduler)
